src/trace/mod.rs

```diff
@@ -19,19 +19,47 @@ pub mod spine_fueled;
 pub use cursor::{Consumer, Cursor, ValueConsumer};
 
 use crate::{
-    algebra::{HasZero, Lattice, MonoidValue},
+    algebra::{HasZero, MonoidValue},
     circuit::Activator,
     time::{AntichainRef, Timestamp},
+    NumEntries,
 };
 use size_of::SizeOf;
+use std::{fmt::Debug, hash::Hash};
+
+/// Trait for data stored in batches.
+///
+/// This trait is used as a bound on `BatchReader::Key` and `BatchReader::Val`
+/// associated types (see [`trait BatchReader`]).  Hence when writing code that
+/// must be generic over any relational data, it is sufficient to impose
+/// `DBData` as a trait bound on types.  Conversely, a trait bound of the form
+/// `B: BatchReader` implies `B::Key: DBData` and `B::Val: DBData`.
+pub trait DBData: Clone + Eq + Ord + Hash + SizeOf + Send + Debug + 'static {}
```

Comment:
What's the `Debug` requirement for?

Reply:
Debugging :)

I changed `impl Display` for all batch types to require `Key: Debug`, `Val: Debug`, `Time: Debug`, `R: Debug` (otherwise it's nearly useless).  This way I will be easily dump the contents of the batch in any operator implementation without adding `Debug` bounds everywhere.